### PR TITLE
Generate ATT&CK navigator layer files and links

### PR DESCRIPTION
**Pull request by brokensound77**
## Issues
None

## Summary

* generates an ATT&CK navigator layer file for all rules coverage
* generates multiple other layer files based on different categories of rule attributes
* files are generated as part of the package build
* created `update-navigator-gist` command to update staged gists, which are hosting the coverage data from `main`
* update `unit tests` workflow to refresh the staged gist files on all successful merges pushed to main
* adds [ATT&CK navigator coverage](https://ela.st/detection-rules-navigator) badge
* add new [ATT&CK-coverage](https://github.com/elastic/detection-rules/blob/3b425a836ce19e80f6a27ccb3c12e7fd4fe40bce/docs/ATT%26CK-coverage.md) page

image

## Details

Navigator layer files are generated as part of package builds and saved in the `extras` directory of the generated files.

<details>

```
releases/8.2/extras
├── 8.2-changelog-entry.md
├── 8.2-consolidated-rules.json
├── 8.2-consolidated-rules.ndjson
├── 8.2-enriched-rules-index-importable.ndjson
├── 8.2-enriched-rules-index-uploadable.ndjson
├── 8.2-summary.txt
├── 8.2-summary.xlsx
├── 8.2.zip
└── navigator_layers
    ├── Elastic-detection-rules-all.json
    ├── Elastic-detection-rules-indexes-auditbeat-*.json
    ├── Elastic-detection-rules-indexes-filebeat-*.json
    ├── Elastic-detection-rules-indexes-logs-*.json
    ├── Elastic-detection-rules-indexes-logs-aws*.json
    ├── Elastic-detection-rules-indexes-logs-azure*.json
    ├── Elastic-detection-rules-indexes-logs-cyberarkpas.json
    ├── Elastic-detection-rules-indexes-logs-endpoint.events.json
    ├── Elastic-detection-rules-indexes-logs-gcp*.json
    ├── Elastic-detection-rules-indexes-logs-google_workspace*.json
    ├── Elastic-detection-rules-indexes-logs-o365*.json
    ├── Elastic-detection-rules-indexes-logs-okta*.json
    ├── Elastic-detection-rules-indexes-logs-system.json
    ├── Elastic-detection-rules-indexes-logs-windows.json
    ├── Elastic-detection-rules-indexes-metrics-*.json
    ├── Elastic-detection-rules-indexes-packetbeat-*.json
    ├── Elastic-detection-rules-indexes-traces-*.json
    ├── Elastic-detection-rules-indexes-winlogbeat-*.json
    ├── Elastic-detection-rules-platforms.json
    ├── Elastic-detection-rules-tags-active-directory.json
    ├── Elastic-detection-rules-tags-application.json
    ├── Elastic-detection-rules-tags-asset-visibility.json
    ├── Elastic-detection-rules-tags-aws.json
    ├── Elastic-detection-rules-tags-azure.json
    ├── Elastic-detection-rules-tags-cloud.json
    ├── Elastic-detection-rules-tags-collection.json
    ├── Elastic-detection-rules-tags-command-and-control.json
    ├── Elastic-detection-rules-tags-communication.json
    ├── Elastic-detection-rules-tags-configuration-audit.json
    ├── Elastic-detection-rules-tags-continuous-monitoring.json
    ├── Elastic-detection-rules-tags-credential-access.json
    ├── Elastic-detection-rules-tags-cyberarkpas.json
    ├── Elastic-detection-rules-tags-data-protection.json
    ├── Elastic-detection-rules-tags-defense-evasion.json
    ├── Elastic-detection-rules-tags-discovery.json
    ├── Elastic-detection-rules-tags-elastic.json
    ├── Elastic-detection-rules-tags-execution.json
    ├── Elastic-detection-rules-tags-gcp.json
    ├── Elastic-detection-rules-tags-google-workspace.json
    ├── Elastic-detection-rules-tags-host.json
    ├── Elastic-detection-rules-tags-identity-and-access.json
    ├── Elastic-detection-rules-tags-identity.json
    ├── Elastic-detection-rules-tags-impact.json
    ├── Elastic-detection-rules-tags-initial-access.json
    ├── Elastic-detection-rules-tags-lateral-movement.json
    ├── Elastic-detection-rules-tags-linux.json
    ├── Elastic-detection-rules-tags-log-auditing.json
    ├── Elastic-detection-rules-tags-macos.json
    ├── Elastic-detection-rules-tags-microsoft-365.json
    ├── Elastic-detection-rules-tags-ml.json
    ├── Elastic-detection-rules-tags-monitoring.json
    ├── Elastic-detection-rules-tags-network-security.json
    ├── Elastic-detection-rules-tags-network.json
    ├── Elastic-detection-rules-tags-okta.json
    ├── Elastic-detection-rules-tags-persistence.json
    ├── Elastic-detection-rules-tags-post-execution.json
    ├── Elastic-detection-rules-tags-privilege-escalation.json
    ├── Elastic-detection-rules-tags-secops.json
    ├── Elastic-detection-rules-tags-threat-detection.json
    ├── Elastic-detection-rules-tags-windows.json
    └── Elastic-detection-rules-tags-zoom.json

```

</details>

The files are stored in a gist and updated whenever `main` has an updated pushed. This is done via the existing github workflow which builds the package.

### CODE REVIEW

File: docs/coverage.md
Comment (brokensound77) [inline docs/coverage.md:19]: some of these are throwing parsing errors from navigator. I will look into it.

<details>

error
```
main-es2018.84172b115fe9ba0742ee.js:formatted:130688 TypeError: Cannot read properties of null (reading 'name')
    at gm.deSerializeDomainVersionID (main-es2018.84172b115fe9ba0742ee.js:formatted:128635:31)
    at Object.next (main-es2018.84172b115fe9ba0742ee.js:formatted:130679:39)
    at u.__tryOrUnsub (main-es2018.84172b115fe9ba0742ee.js:formatted:24047:23)
    at u.next (main-es2018.84172b115fe9ba0742ee.js:formatted:24010:157)
    at l._next (main-es2018.84172b115fe9ba0742ee.js:formatted:23970:34)
    at l.next (main-es2018.84172b115fe9ba0742ee.js:formatted:23955:40)
    at s._next (main-es2018.84172b115fe9ba0742ee.js:formatted:84031:34)
    at s.next (main-es2018.84172b115fe9ba0742ee.js:formatted:23955:40)
    at s._next (main-es2018.84172b115fe9ba0742ee.js:formatted:89949:39)
    at s.next (main-es2018.84172b115fe9ba0742ee.js:formatted:23955:40)
next @ main-es2018.84172b115fe9ba0742ee.js:formatted:130688
```

source
```javascript
async loadLayerFromURL(e, t, n=!1) {
                return new Promise(async(r,i)=>{
                    let o = this.http.get(e).subscribe({
                        next: async i=>{
                            let o = this.viewModelsService.newViewModel("loading layer...", void 0);
                            try {
                                if (o.deSerializeDomainVersionID(i),
                                !this.dataService.getDomain(o.domainVersionID))
                                    throw {
                                        message: "Error: '" + o.domain + "' (" + o.version + ") is an invalid domain."
                                    };
                                await this.layerUpgrade(o, i, t, n),
                                console.log("loaded layer from", e),
                                r(null)
                            } catch (s) {
                                console.error(s),
                                this.viewModelsService.destroyViewModel(o),
                                alert("ERROR parsing layer from " + e + ", check the javascript console for more information."),
                                r(null)
                            }
```

</details>

File: detection_rules/ghwrap.py
Comment (brokensound77) [inline detection_rules/ghwrap.py:122]:
```suggestion
        # retrieve all existing file names which are not in the file_map and overwrite them to empty to delete files
```

**Review by Mikaayenson — COMMENTED** [review]
I didn't see any major logic issues. It might be a good idea to add docstrings.

File: detection_rules/devtools.py
Comment (brokensound77) [reply detection_rules/devtools.py:749]: The command itself is written to be generic and update _any_ gist with _any_ directory of navigator json files. We defer to click to handle a lot of that communication, so if a token is missing, it will prompt a user. This includes defaults, meaning if you have a non default, you as a caller should already know the values.

So all of the `Notes` would be overly specific to that one use case.

The defaults are set as they are to simplify calling from the github actions.

File: detection_rules/devtools.py
Comment (brokensound77) [reply detection_rules/devtools.py:744]: Looks like it passes a 401 which should be descriptive enough, especially since this is an internal only function (under `dev`)

```
401 - {"message":"Bad credentials","documentation_url":"https://docs.github.com/rest"}
```

File: detection_rules/devtools.py
Comment (brokensound77) [reply detection_rules/devtools.py:43]: This is the default staged gist which is hosting the files for the default navigator links. Automation from github actions will have access to update it

File: .github/workflows/pythonpackage.yml
Comment (brokensound77) [reply .github/workflows/pythonpackage.yml:54]: `GITHUB_TOKEN` is a special environment variable parsed by the CLI https://github.com/elastic/detection-rules/blob/0aeb7399d448186dd01fdb7d7b6d6b776c91e462/detection_rules/devtools.py#L44-L51

File: detection_rules/packaging.py
Comment (brokensound77) [reply detection_rules/packaging.py:80]: py3+ doesn't require it. Better to update flake8 config if it is an issue

File: docs/ATT&CK-coverage.md
Comment (brokensound77) [reply docs/ATT&CK-coverage.md:22]: The table should stay consistent unless the `NavigatorBuilder` is updated, since permalinks are being used

File: detection_rules/packaging.py
Comment (brokensound77) [reply detection_rules/packaging.py:77]: I think it makes more sense here to be globally available and within the best context

File: detection_rules/navigator.py
Comment (brokensound77) [reply detection_rules/navigator.py:22]: That is alphabetized 🙂. Also, it is the default arrangement by the ATT&CK object

File: detection_rules/ghwrap.py
Comment (brokensound77) [reply detection_rules/ghwrap.py:129]: I think 401 for bad creds. Do you think 404 not found is not specific enough for `a gist file that they can not write to.`. GitHub is sometimes overly generic and so we'd have to make assumptions, because it could be a bad gist ID just as much as missing, so in this case, I think we should be ok (especially for an internal `dev` command)

File: detection_rules/navigator.py
Comment (brokensound77) [reply detection_rules/navigator.py:271]: I'm not sure I am following, can you elaborate? Maybe the line breaks of the args?

File: detection_rules/navigator.py
Comment (brokensound77) [reply detection_rules/navigator.py:22]: Oh sorry, I see now. Yea we can make that change :+1:

File: detection_rules/devtools.py
Comment (Mikaayenson) [reply detection_rules/devtools.py:744]: The error I received was a 404, which may have occurred when trying to use your gist_id. The error went away after using my own gist_id.

File: detection_rules/ghwrap.py
Comment (Mikaayenson) [reply detection_rules/ghwrap.py:129]: Both may be good. However, I only received the 404 error when testing.

File: detection_rules/navigator.py
Comment (Mikaayenson) [reply detection_rules/navigator.py:271]: This is one of those cosmetic comments for consistency. The line length for most methods/functions appear to be 120.  We should probably decide to visually have multiple lines for the function signature or the minimum number of lines based on the preferred line length.

It's all really personal preference. I just noticed that it was slightly different. 

```
    def save_layer(self, layer_name: str, directory: Path, 
                   layer_key: Optional[str] = None, verbose=True) -> (Path, dict):

```

vs

```
    def save_layer(self,
                   layer_name: str,
                   directory: Path,
                   layer_key: Optional[str] = None,
                   verbose=True
                   ) -> (Path, dict):
```

File: .github/workflows/pythonpackage.yml
Comment (Mikaayenson) [reply .github/workflows/pythonpackage.yml:54]: This suggestion would add the NAVIGATOR_GIST_ID token, not remove the `GITHUB_TOKEN`.

File: docs/ATT&CK-coverage.md
Comment (Mikaayenson) [reply docs/ATT&CK-coverage.md:22]: Got it. So will the links replace the `brokensound77` to the main elastic repo?

File: detection_rules/packaging.py
Comment (Mikaayenson) [reply detection_rules/packaging.py:77]: Got it. so `packaging.py` is more of a global file.

File: detection_rules/devtools.py
Comment (brokensound77) [reply detection_rules/devtools.py:744]: ah, so 404 if authed but no gist_id associated with that user. We can catch this and suggest to validate gist_id exists and user has access 👍

File: detection_rules/devtools.py
Comment (brokensound77) [reply detection_rules/devtools.py:749]: do you use `gh` CLI tool? We parse that config if it exists as part of the process

https://github.com/elastic/detection-rules/blob/0aeb7399d448186dd01fdb7d7b6d6b776c91e462/detection_rules/devtools.py#L48-L49

File: .github/workflows/pythonpackage.yml
Comment (brokensound77) [reply .github/workflows/pythonpackage.yml:54]: ah I see what you are saying - I thought about it, but the gist URL is in the navigator links, which then includes the gist_id, so it is public and so no need to store it as a secret IMO.

File: docs/ATT&CK-coverage.md
Comment (brokensound77) [reply docs/ATT&CK-coverage.md:22]: if we changed default gist_id, they would change, yea

File: detection_rules/devtools.py
Comment (brokensound77) [reply detection_rules/devtools.py:745]: cd36a83

File: detection_rules/navigator.py
Comment (brokensound77) [reply detection_rules/navigator.py:22]: cd36a83

File: detection_rules/navigator.py
Comment (brokensound77) [reply detection_rules/navigator.py:271]: So I had always preferred your first option in this situation, but I have since leaning towards option B as of late because there have been more occurrences of odd breaks, mostly due to type hinting (also to stay more consistent with the stylings of other languages).

For now, I don't think we need to get too caught up on anything stylistic (though style guide is important for some cases)

File: detection_rules/ghwrap.py
Comment (brokensound77) [reply detection_rules/ghwrap.py:128]: d36a83

File: detection_rules/packaging.py
Comment (brokensound77) [reply detection_rules/packaging.py:366]: because most of the documentation (and typing, defaults, etc) can be derived from existing components, if we feel the need for docs, we could leverage that route. There are som existing libs :
https://pypi.org/project/mkdocs-click/ 
https://sphinx-click.readthedocs.io/en/latest/usage/

or we could roll our own.

We should punt that to a dedicated issue though

**Comment by brokensound77** [comment]
After merging #1791, I manually refreshed the gists

image

**Review by Mikaayenson — APPROVED** [review]
LGTM

**Review by w0rk3r — APPROVED** [review]
LGTM

**Pull request merged**
